### PR TITLE
TE-502 Creating Fields from ScreenBuffer for output.

### PR DESCRIPTION
**Pull request by kbayraktar**

### CODE REVIEW

File: src/main/java/org/testeditor/fixture/host/HostDriverFixture.java [hunk 1]
```diff
@@ -362,6 +365,48 @@ public Result moveCursor(String elementLocator, LocatorStrategy locatorType) {
...
+     * <strong>Caution:</strong> But before it will be verified, if the result
+     * of ScreenBuffer is formatted. If the application program uses field
+     * attributes to define fields on the screen, then the screen is formatted.
+     * If there are no fields defined on the screen, then the screen is
+     * unformatted, and the operator uses it in free-form manner.
+     * 
+     * @return The String reprentation of all found {@link Field}s in
+     *         ScreenBuffer.
+     */
+    public String printFieldsOnScreen() {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/HostDriverFixture.java]: why is it named printFieldsOnScreen? could be buildAllFieldsAsString?

File: src/main/java/org/testeditor/fixture/host/net/Connection.java [hunk 2]
```diff
@@ -214,7 +214,7 @@ public Result doCommand(final String command) {
             List<String> lines = readOutput();
             int size = lines.size();
             if (size > 0) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/net/Connection.java]: check should be size > 1, if less some error should be written

File: src/main/java/org/testeditor/fixture/host/s3270/Result.java [hunk 3]
```diff
@@ -25,12 +25,38 @@
...
+     *            "U F U C(abcdefg.google.mainframe.de) I 2 24 80 3 9 0x0 -"
+     * @param resultOfCommand
+     *            String "ok" for command was executed succesfully, "error"
+     *            otherwise.
+     */
+    public Result(final List<String> dataLines, final String statusString, String resultOfCommand) {
         this.dataLines = dataLines;
         this.statusString = statusString;
+        this.resultOfCommand = resultOfCommand;
+        this.status = new Status(statusString);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/s3270/Result.java:59]: public void createStatus can be removed, now.
do you still need setStatus?

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+     *            Raw String of the whole line in the format: "data: SF(c0=e9)
+     *            7a 2f 4f 53 SF(c0=c0) 20 5a 31 38 20 4c 65 76 65 6c SF(c0=d1)
+     *            20 30 36 30 39"
+     * @return The complete List of all rawField String representations in one
+     *         row. String reprentation of a {@link Field} in the format:
+     *         "(c0=c1) 33 34 35", "(c0=c2) 36 37 38", ...
+     */
+    private ArrayList<String> separateFields(String rowAsHex) {
+        ArrayList<String> fieldValues = new ArrayList<String>();
+        String dataReplaced = rowAsHex.replace("data: ", "");
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: 'data: ' should be removed only at the start of the string! replace removes all occurances!

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+     * </ul>
+     * <b>Caution:</b> the result has to be delivered in formatted structure
+     * from mainframe, otherwise we can not parse any field information. See
+     * also
+     * {@link org.testeditor.fixture.host.HostDriverFixture#printFieldsOnScreen()}
+     * 
+     * @see org.testeditor.fixture.host.screen.HostTestSite#createHostScreenAsHex()
+     * @return All found {@link Field} on screen as {@link ArrayList}.
+     */
+    private ArrayList<ArrayList<Field>> createFields(List<String> screenAsHex) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: Why do I need an array of an array, why not simply a list of fields?

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        ArrayList<ArrayList<Field>> allFieldsOnScreen = createFields(screenAsHex);
+        for (ArrayList<Field> line : allFieldsOnScreen) {
+
+            for (Field field : line) {
+                String numeric = "";
+                // only Input Fields can be set as numeric fields.
+                if (!field.isProtected()) {
+                    numeric = String.valueOf(field.IsNumeric());
+                }
+                rowCounter++;
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: this is a fieldCounter, not a rowCounter, right?

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+     *            7a 2f 4f 53 SF(c0=c0) 20 5a 31 38 20 4c 65 76 65 6c SF(c0=d1)
+     *            20 30 36 30 39"
+     * @return {@link Field}
+     */
+    private Field parseField(ArrayList<String> row, String rawField, int rowCount, int fieldCount,
+            String completeRowAsHex) {
+        // like separated "c0=c1", "33", "34", "35"
+        String[] splittedFieldAttributes = rawField.split(" ");
+        Field field = new Field();
+        field.setFieldAttributes(row, fieldCount);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java:204]: Why do I need the whole row here?

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+     *            The row number on whole screen.
+     * @param fieldCount
+     *            The number of a field in one line.
+     * @param completeRowAsHex
+     *            Raw String of the whole line in the format: "data: SF(c0=e9)
+     *            7a 2f 4f 53 SF(c0=c0) 20 5a 31 38 20 4c 65 76 65 6c SF(c0=d1)
+     *            20 30 36 30 39"
+     * @return {@link Field}
+     */
+    private Field parseField(ArrayList<String> row, String rawField, int rowCount, int fieldCount,
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java:199]: why do I need the whole row?

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+     * 
+     * @param screenAsHex
+     *            Input is the result of x3270 "Readbuffer(Ascii) command. <br>
+     *            <b>Caution:</b> the result has to be delivered in formatted
+     *            structure from mainframe, otherwise we can not parse any field
+     *            information.
+     * @see org.testeditor.fixture.host.screen.HostTestSite#createHostScreenAsHex()
+     * @return the field reprentations on screen as String
+     */
+    public ArrayList<Field> getAllFields(List<String> screenAsHex) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: if createFields would return a list already, this method would be obsolete

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 10]
```diff
@@ -115,23 +127,23 @@ public String printAllFieldAsString(List<String> screenAsHex) {
...
-                    numeric = String.valueOf(field.IsNumeric());
-                }
-                rowCounter++;
-                Formatter formatter = new Formatter();
-                formatter.format(outputFormat, rowCounter, field.getRowCount(), field.getFieldCountPerline(),
-                        field.getCharacterWidth(), field.getCharacterStartpoint(), field.getEndpoint(), numeric,
-                        field.isProtected(), field.isHidden(), " '" + field.getStringReperentation() + "'");
-                screenBuffer.append(formatter + "\n");
-                formatter.close();
+            // for (Field field : line) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: get rid of comment

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 11]
```diff
@@ -115,23 +127,23 @@ public String printAllFieldAsString(List<String> screenAsHex) {
...
+                numeric = String.valueOf(field.IsNumeric());
             }
+            fieldCounter++;
+            Formatter formatter = new Formatter();
+            formatter.format(outputFormat, fieldCounter, field.getRowCount(), field.getFieldCountPerline(),
+                    field.getCharacterWidth(), field.getCharacterStartpoint(), field.getEndpoint(), numeric,
+                    field.isProtected(), field.isHidden(), " '" + field.getStringReperentation() + "'");
+            screenBuffer.append(formatter + "\n");
+            formatter.close();
+            // }
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: get rid of comment

File: src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java [hunk 12]
```diff
@@ -149,16 +161,17 @@ public String printAllFieldAsString(List<String> screenAsHex) {
...
-        ArrayList<Field> allFields = new ArrayList<Field>();
-        ArrayList<ArrayList<Field>> allFieldsOnScreen = createFields(screenAsHex);
-        for (ArrayList<Field> line : allFieldsOnScreen) {
-            for (Field field : line) {
-                allFields.add(field);
-            }
-        }
-        return allFields;
-    }
+    // public ArrayList<Field> getAllFields(List<String> screenAsHex) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/screen/TerminalScreen.java]: get rid of comment

File: src/test/java/org/testeditor/fixture/host/net/ConnectionTest.java [hunk 13]
```diff
@@ -43,4 +43,24 @@ public void testLinkedListSpyCorrect() throws IOException {
         assertEquals("ok", actualResult.getResultOfCommand());
     }
 
+    @Test
+    public void testConnectionWithoutData() throws IOException {
+        //
```
Comment (gunther-bachmann) [inline src/test/java/org/testeditor/fixture/host/net/ConnectionTest.java]: add given, when, then perhaps?

File: src/main/java/org/testeditor/fixture/host/HostDriverFixture.java [hunk 14]
```diff
@@ -390,15 +390,17 @@ public Result moveCursor(String elementLocator, LocatorStrategy locatorType) {
      * @return The String reprentation of all found {@link Field}s in
      *         ScreenBuffer.
      */
-    public String printFieldsOnScreen() {
+    @FixtureMethod
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/HostDriverFixture.java]: why is this a fixture method?

File: src/main/java/org/testeditor/fixture/host/HostDriverFixture.aml [hunk 15]
```diff
@@ -29,6 +29,11 @@ interaction type typeAt {
     method = HostDriverFixture.typeAt(element, locatorStrategy, value)
 }
 
+interaction type printFieldsOnLog {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/host/HostDriverFixture.aml:32]: what do I need this for?

**Comment by franzbecker** [comment]
3k LoC is too much to review, I'll pass

File: src/main/java/org/testeditor/fixture/host/s3270/Result.java (hunk 3 above)
Comment (kbayraktar) [reply src/main/java/org/testeditor/fixture/host/s3270/Result.java:59]: done :thumbsup:

**Pull request merged**